aten/src/ATen/core/jit_type.h

```diff
@@ -0,0 +1,923 @@
+#pragma once
+
+#include <ATen/core/ivalue.h>
+#include <ATen/core/interned_strings.h>
+#include <ATen/core/functional.h>
+#include <ATen/core/Type.h>
+#include <ATen/core/TensorMethods.h>
+
+#include <caffe2/core/common.h>
+
+#include <memory>
+#include <iostream>
+#include <type_traits>
+
+namespace c10 {
+
+#define C10_FORALL_TYPES(_) \
+_(DynamicType) \
+_(TensorType) \
+_(CompleteTensorType) \
+_(UndefinedTensorType) \
+_(TupleType) \
+_(ListType) \
+_(NumberType) \
+_(FloatType) \
+_(FutureType) \
+_(IntType) \
+_(NoneType) \
+_(StringType) \
+_(GeneratorType) \
+_(BoolType) \
+_(OptionalType) \
+_(VarType) \
+_(WorldType)
+
+enum class TypeKind {
+#define DEFINE_TYPE(T) T,
+  C10_FORALL_TYPES(DEFINE_TYPE)
+#undef DEFINE_TYPE
+};
+
+struct Type;
+using TypePtr = std::shared_ptr<Type>;
+
+struct CAFFE2_API Type : std::enable_shared_from_this<Type> {
+private:
+  TypeKind kind_;
+  template<typename T>
+  static std::shared_ptr<T> sliceType(std::shared_ptr<const T> ptr) {
+    auto result = std::make_shared<typename std::remove_const<T>::type>(*ptr);
+    // XXX: the line above will correctly slice the struct, and make its runtype
+    // type exactly equal to T. However, kind_ is a field of Type, so it will simply
+    // be copied, and we need to fix it in here to match the dynamic type.
+    result->kind_ = T::Kind;
+    return result;
+  }
+
+protected:
+  Type(TypeKind kind)
+    : kind_(kind) {}
+
+public:
+  virtual bool operator==(const Type& rhs) const = 0;
+
+  // subtyping relation. By default, we return true for the case
+  // when the type is exactly equal
+  virtual bool isSubtypeOf(const TypePtr rhs) const {
+    return *this == *rhs;
+  }
+
+  // How this type will appear in FunctionSchema declarations
+  virtual std::string str() const = 0;
+
+  // How this type will appear as if it were a type annotation in Python
+  // which is sometimes different than how it appears in declarations (e.g. int[] vs List[int])
+  virtual std::string python_str() const {
+    return str();
+  }
+
+  TypeKind kind() const {
+    return kind_;
+  }
+
+  virtual bool requires_grad() const { return false; }
+
+  // Dynamically cast this object to the subclass indicated by the
+  // template variable, returning nullptr if the cast is invalid.
+  // NOTE: if the cast succeeds, but the casted kind is not the
+  // run-time kind of the type, we also slice the structure, so
+  // that assignments of those types to values don't accidentally
+  // inherit more detailed information from subclasses.
+  template<typename T>
+  std::shared_ptr<T> cast() {
+    auto r = caffe2::dynamic_pointer_cast_if_rtti<T>(shared_from_this());
```

Comment:
1) make sense, I'll fix that
2) Should I just wait until the code migrates away from using RTTI at all?  Is that planned? @zdevito

Reply:
I am not aware of any concrete plans to remove RTTI, though it should be possible to do for this case if absolutely necessary since the type hierarchy is tagged.

Reply:
It doesn't. You would have to implement the subtyping check based on tags.